python/src/vmaf/core/noref_feature_extractor.py

```diff
@@ -431,3 +431,67 @@ def mscn_extract_niqe(cls, img, patch_size, mode):
             assert False
 
         return list_features
+
+
+class SiTiNorefFeatureExtractor(NorefExecutorMixin, FeatureExtractor):
+
+    TYPE = "SITI_noref_feature"
+    VERSION = "1.0"
+
+    ATOM_FEATURES = ['si', 'ti']  # order matters
+
+    @staticmethod
+    def sobel_filt(img):
+
+        dx = ndimage.sobel(img, 1)  # horizontal derivative
+        dy = ndimage.sobel(img, 0)  # vertical derivative
+        mag = np.hypot(dx, dy)  # magnitude
+
+        return mag
+
+    def _generate_result(self, asset):
+        # routine to generate feature scores in the log file.
+
+        quality_w, quality_h = asset.quality_width_height
+        with YuvReader(filepath=asset.dis_workfile_path, width=quality_w,
+                       height=quality_h,
+                       yuv_type=self._get_workfile_yuv_type(asset)) \
+                as dis_yuv_reader:
+            scores_mtx_list = []
+            i = 0
+            for dis_yuv in dis_yuv_reader:
+                dis_y = dis_yuv[0].astype('int32')
+                mag = self.sobel_filt(dis_y)
+                si = np.std(mag)
+                if i == 0:
+                    ti = 0
+                else:
+                    ti = np.std(dis_y - dis_y_prev)
+                dis_y_prev = copy.deepcopy(dis_y)
+                scores_mtx_list.append(np.hstack(([si], [ti])))
+                i += 1
+            scores_mtx = np.vstack(scores_mtx_list)
+
+            # write scores_mtx to log file
+        log_file_path = self._get_log_file_path(asset)
+        with open(log_file_path, "wb") as log_file:
+            np.save(log_file, scores_mtx)
+
+    def _get_feature_scores(self, asset):
```

Comment:
The _get_feature_scores() method seems to be used in a number of other places in the file, maybe we can find a way to re-use existing logic?

Reply:
Yes. it is generic and we can move it to the abstract class as the default (it is always overridable).

Reply:
Hmm. i looked into refactoring the code. The base class already have a _get_feature_scores method. Perhaps I'll keep it this way for now, and look for a more systematic way of cleaning up the code in the near future.